File: Chroma-Core/src/main/java/buttondevteam/lib/chat/CoreCommandBuilder.java
package buttondevteam.lib.chat;

import buttondevteam.lib.chat.commands.CommandArgument;
import buttondevteam.lib.chat.commands.SubcommandData;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.tree.CommandNode;

import java.util.Map;
import java.util.function.Function;

public class CoreCommandBuilder<S extends Command2Sender, TC extends ICommand2<?>> extends LiteralArgumentBuilder<S> {
	private final SubcommandData.SubcommandDataBuilder<TC, S> dataBuilder;

	protected CoreCommandBuilder(String literal, Class<?> senderType, Map<String, CommandArgument> arguments, CommandArgument[] argumentsInOrder, TC command) {
		super(literal);
		dataBuilder = SubcommandData.<TC, S>builder().senderType(senderType).arguments(arguments)
			.argumentsInOrder(argumentsInOrder).command(command);
	}

	@Override
	protected CoreCommandBuilder<S, TC> getThis() {
		return this;
	}

	public static <S extends Command2Sender, TC extends ICommand2<?>> CoreCommandBuilder<S, TC> literal(String name, Class<?> senderType, Map<String, CommandArgument> arguments, CommandArgument[] argumentsInOrder, TC command) {
		return new CoreCommandBuilder<>(name, senderType, arguments, argumentsInOrder, command);
	}

	public static <S extends Command2Sender, TC extends ICommand2<?>> CoreCommandBuilder<S, TC> literalNoOp(String name) {
		return literal(name, Command2Sender.class, Map.of(), new CommandArgument[0], null);
	}

	/**
	 * Static help text added through annotations. May be overwritten with the getter.
	 *
	 * @param helpText Help text shown to the user
	 * @return This instance
	 */
	public CoreCommandBuilder<S, TC> helps(String[] helpText) {
		dataBuilder.staticHelpText(helpText);
		return this;
	}

	/**
	 * Custom help text that depends on the context. Overwrites the static one.
	 * The function receives the sender but its type is not guaranteed to match the one at the subcommand.
	 * It will either match or be a Command2Sender, however.
	 *
	 * @param getter The getter function receiving the sender and returning the help text
	 * @return This instance
	 */
	public CoreCommandBuilder<S, TC> helps(Function<Object, String[]> getter) {
		dataBuilder.helpTextGetter(getter);
		return this;
	}

	public CoreCommandBuilder<S, TC> permits(Function<S, Boolean> permChecker) {
		dataBuilder.hasPermission(permChecker);
		return this;
	}

	@Override
	public CoreCommandNode<S, TC> build() {
		var result = new CoreCommandNode<S, TC>(this.getLiteral(), this.getCommand(), this.getRequirement(),
			this.getRedirect(), this.getRedirectModifier(), this.isFork(),
			dataBuilder.build());

		for (CommandNode<S> node : this.getArguments()) {
			result.addChild(node);
		}

		return result;
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/chat/CoreCommandBuilder.kt
package buttondevteam.lib.chat

import buttondevteam.lib.chat.commands.CommandArgument
import buttondevteam.lib.chat.commands.CommandUtils.coreArgument
import buttondevteam.lib.chat.commands.NoOpSubcommandData
import buttondevteam.lib.chat.commands.SubcommandData
import com.mojang.brigadier.builder.ArgumentBuilder
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import java.lang.reflect.Method

class CoreCommandBuilder<S : Command2Sender, TC : ICommand2<S>, TSD : NoOpSubcommandData> private constructor(
    literal: String,
    val data: TSD
) : LiteralArgumentBuilder<S>(literal) {

    override fun getThis(): CoreCommandBuilder<S, TC, TSD> {
        return this
    }

    override fun build(): CoreCommandNode<S, TSD> {
        val result = CoreCommandNode<_, _>(
            literal,
            command,
            requirement,
            this.redirect,
            this.redirectModifier,
            this.isFork,
            data
        )
        for (node in arguments) {
            result.addChild(node)
        }
        return result
    }

    override fun then(argument: ArgumentBuilder<S, *>): LiteralArgumentBuilder<S> {
        super.then(argument)
        if (data is SubcommandData<*, *>) {
            @Suppress("UNCHECKED_CAST")
            arguments.forEach { it.coreArgument()?.commandData = data as SubcommandData<*, S> }
        }
        return this
    }

    companion object {
        /**
         * Start building an executable command node.
         *
         * @param name The subcommand name as written by the user
         * @param senderType The expected command sender type based on the subcommand method
         * @param arguments A map of the command arguments with their names as keys
         * @param argumentsInOrder A list of the command arguments in the order they are expected
         * @param command The command object that has this subcommand
         * @param helpTextGetter Custom help text that can depend on the context. The function receives the sender as the command itself receives it.
         * @param hasPermission A function that determines whether the user has permission to run this subcommand
         * @param annotations All annotations implemented by the method that executes the command
         * @param fullPath The full command path of this subcommand.
         */
        fun <S : Command2Sender, TC : ICommand2<S>> literal(
            name: String,
            senderType: Class<*>,
            arguments: Map<String, CommandArgument>,
            argumentsInOrder: List<CommandArgument>,
            command: TC,
            helpTextGetter: (Any) -> Array<String>,
            hasPermission: (S, SubcommandData<TC, S>) -> Boolean,
            annotations: Array<Annotation>,
            fullPath: String,
            method: Method
        ): CoreCommandBuilder<S, TC, SubcommandData<TC, S>> {
            return CoreCommandBuilder(
                name,
                SubcommandData(senderType, arguments, argumentsInOrder, command, helpTextGetter, hasPermission, annotations, fullPath, method)
            )
        }

        /**
         * Start building a no-op command node.
         *
         * @param name The subcommand name as written by the user
         * @param helpTextGetter Custom help text that can depend on the context. The function receives the sender as the command itself receives it.
         */
        fun <S : Command2Sender, TC : ICommand2<S>> literalNoOp(
            name: String,
            helpTextGetter: (Any) -> Array<String>,
        ): CoreCommandBuilder<S, TC, NoOpSubcommandData> {
            return CoreCommandBuilder(name, NoOpSubcommandData(helpTextGetter))
        }
    }
}